src/projectDB/projectdatabase.cpp

```diff
@@ -6,11 +6,45 @@
 #include "connection.h"
 #include "cursor.h"
 #include "mzSample.h"
+#include "projectversioning.h"
 #include "schema.h"
 
-ProjectDatabase::ProjectDatabase(const string& dbFilename)
+ProjectDatabase::ProjectDatabase(const string& dbFilename,
+                                 const string& version)
 {
     _connection = new Connection(dbFilename);
+
+    // figure out whether this database needs upgrade
+    using namespace ProjectVersioning;
+    auto aheadBy = 0;
+    auto versionInfo = extractVersionInfoFromTag(version, &aheadBy);
+    auto appVersion = Version(versionInfo.first);
+    auto requiredDbVersion = getDbVersionForApp(appVersion);
+    auto currentDbVersion = this->version();
+
+    // if build is ahead by non-zero number of commits, then its a dev build
+    if (aheadBy > 0) {
+        requiredDbVersion = getLatestDbVersion();
+    }
+
+    if (currentDbVersion != requiredDbVersion && !this->isEmpty()) {
+        // upgrade needed, close existing connection
+        delete _connection;
+
+        string upgradeScript = generateUpgradeScript(currentDbVersion,
+                                                     requiredDbVersion);
+        cout << "Debug: Upgrading database from v"
+             << currentDbVersion
+             << " to v"
+             << requiredDbVersion
+             << endl;
+        upgradeDatabase(dbFilename, upgradeScript, version);
```

Comment:
Call only if required. There's no check here or in `upgradeDatabase`

Reply:
If control has reached this point, then it definitely does need an upgrade (and its a logical error if it does not). The check on [line 30](https://github.com/ElucidataInc/ElMaven/pull/972/files#diff-b65d4507eb26f772cc8ef1fef3e8c8b9R30) is the only check necessary.

Reply:
I was talking about the case where the `upgradeScript` might be empty.  From your point it seems as if, `generateUpgradeScript` will never return an empty string

Reply:
It might. When there is only a single table being added in the entire release. In that case an upgrade you are correct. An unnecessary upgrade should not be performed. I'll add a check in `upgradeDatabase` function, because we should still back-up user's original database and that piece of code should still be executed. Let me know if you disagree.

Reply:
Makes sense, thanks!